docs/rfcs/0002-browser-search-rewrite.md

```diff
@@ -0,0 +1,127 @@
+
+Using the `BrowserStore` for state will make the app use asynchronous patterns to observe the state and with that avoid blocking the main thread, as well as force the app to deal with the initial state of having no (default) `SearchEngine` yet.
+
+Since the state will move to `browser-state`, additional functionality (querying suggestions, storage, middleware) can move to `feature-search`. This will make `browser-search` obsolete. With that introducing a `concept-search` is not required. An app can use a custom implementation by replacing the middleware, storage or handling search state completely differently.
+
+## Reference-level explanation
+
+The new implementation will use `browser-state` to model all search state.
+
+Functionality on top of the state (querying suggestions, storage, middleware) will live in `feature-search`. With that `browser-search` will no longer be needed and can stay around as-is until no longer used by any app.
+
+### State
+
+`BrowserState` will get a new `search` property with `SearchState` type that lives in `browser-state`:
+
+```Kotlin
+data class BrowserState(
+   // ...
+   val search: SearchState
+)
+
+/**
+ * Value type that represents the state of available search engines.
+ *
+ * @property searchEngines List of loaded search engines.
+ * @property defaultSearchEngineId ID of the default search engine.
+ */
+data class SearchState(
+  val searchEngines: List<SearchEngine>,
+  val defaultSearchEngineId: String
+)
+```
+
+`SearchState` will contain all `SearchEngine`s (bundled and custom). Additional extension methods will make it easier to select specific subsets or the default `SearchEngine`.
+
+`SearchEngine` will be turned into a data class and moved to `browser-store`. Other methods like `buildSearchUrl()` will be implemented as extension methods.
+
+### Storage
+
+There will be two storage classes:
+
+* `SearchEngineStorage`: A persistent storage for `SearchEngine`s. The primary use case is for persisting custom search engines added by users.
+* `SearchEngineDefaults`: A provider of a list of default search engines (for the user's region) based on `list.json` and the search plugins currently included in `browser-search`.
+
+Those storage classes will have visibility `internal` and will not be used by the app directly.
+
+All storage access methods will be suspending functions to avoid thread-blocking access:
+
+```Kotlin
+internal class SearchEngineStorage {
+    suspend fun getSearchEngines() = withContext(Dispatchers.IO) {
+      // ...
+    }
+}
+```
+
+### Middleware
+
+A `SearchMiddleware` that will be installed on `BrowserStore` will be responsible for:
```

Comment:
What about the name `SearchManager`? I think we have manager classes in a-c with similar responsibilities as below.

Reply:
I think using `*Middleware` is more in line with other `*Middleware` that we have implemented now and it describes its purpose (to be added to the store as a middleware).
Using `*Manager` too vague? And it doesn't really "manage" the state anymore.

Some thing using "manager" is an anti pattern:
* http://www.bright-green.com/blog/2003_02_25/naming_java_classes_without_a.html
* https://blog.codinghorror.com/i-shall-call-it-somethingmanager/